Drop wrappers from Internal Server

maneeshpm opened the pull request:
This is the first step, fixes #430 

The aim of this step is to internally remove usage of wrapper structures in libkiwix as much as possible. A majority of work in this step will be focused around bringing this change in the internal server which will in turn add new methods/change existing code to use libzim structures such as `zim::Archive`, `zim::Searcher`, `zim::SearchResultSet` etc directly instead of adding additional layer of `kiwix::Reader` or `kiwix::Searcher`.

Changes included in this PR:
- Drop wrappers from `InternalServer` completely
- Add libzim structure mirror members to `Library` and other kiwix structures
- Add new method in `SearchRenderer` to use libzim structures directly
- Move helpers for handling archive to `archiveTools`

maneeshpm commented on include/library.h:
```diff
@@ -146,6 +147,7 @@ class Library
 {
   std::map<std::string, kiwix::Book> m_books;
   std::map<std::string, std::shared_ptr<Reader>> m_readers;
+  std::map<std::string, std::shared_ptr<zim::Archive>> m_archives;
```
@mgautierfr Even though all the tests are passing, the addition of this line causes a segmentation fault when I try to run `kiwix-serve` :confused: Can you help figure it out? I can confirm the issue is isolated to this one line irrespective of other changes.

kelson42 replied:
@maneeshpm Would be good to extend the tests, so the CI fails as well.

maneeshpm replied:
@kelson42 Good idea! Let me try that, maybe it will reveal the underlying issue.

maneeshpm replied:
@kelson42 I got this issue because, for some reason, I had to rebuild kiwix-tools again against the modified libkiwix even though no API has been changed. I guess it's an isolated incident with my system. Closing this discussion.

kelson42 replied:
good!

mgautierfr replied:
This is what we call a ABI break.
If you add a member to a structure, the size of this structure is changed.
So all the code of kiwix-tools compiled with previous version is using the old size of library but the new library is using the new size. So the new libkiwix writes data where it should not (where kiwix-tools has not reserved space for it).

maneeshpm replied:
Aaah cool. The segmentation fault makes sense now, Thanks!

kelson42 commented:
@maneeshpm Thx for this important PR!

maneeshpm commented on include/library.h:
```diff
@@ -146,6 +147,7 @@ class Library
 {
   std::map<std::string, kiwix::Book> m_books;
   std::map<std::string, std::shared_ptr<Reader>> m_readers;
+  std::map<std::string, std::shared_ptr<zim::Archive>> m_archives;
```
@mgautierfr Currently the archive is held as a unique pointer inside the reader, do you think for the time being making it a shared pointer and copy into the above m_archive map when creating instead of directly making it from the path is a good idea?

mgautierfr replied:
Yes, we should move to a shared_ptr.
I think the "correct" API would be to always creating a reader from a existing archive.
For compatibility issue, we should have a constructor taking a path an create the archive internally.
On the library side, we should also first create the archive and create the Reader using it.

maneeshpm commented on src/server/internalServer.cpp:
```diff
@@ -328,46 +329,119 @@ std::unique_ptr<Response> InternalServer::build_homepage(const RequestContext& r
+  return false;
+}
+
+bool getMetadata(const zim::Archive* const archive, const string& name, string& value)
+{
+  try {
+    value = archive->getMetadata(name);
+    return true;
+  } catch(zim::EntryNotFound& e) {
+    return false;
+  }
+}
```
I wanted to keep the code same as that used in reader and change it in a later commit. I probably left it out, will do it.

maneeshpm commented on src/server/internalServer.cpp:
```diff
@@ -505,10 +548,10 @@ std::unique_ptr<Response> InternalServer::handle_suggest(const RequestContext& r
 
   /* Propose the fulltext search if possible */
-  if (reader->hasFulltextIndex()) {
+  if (archive->hasFulltextIndex()) {
```
Aah right! It can be a nullptr. This block should be executed only if we are not dealing with a nullptr, and if it is, we should be returning a 404(like we do in handle_search). Perhaps we should check if the archive is a nullptr and return a 404 in such a case before we actually start handling the search.

mgautierfr replied:
Yes, test and early return is good here.

maneeshpm commented on include/search_renderer.h:
```diff
@@ -40,6 +41,7 @@ class SearchRenderer
    *                          Used to generate pagination links.
    */
   SearchRenderer(Searcher* searcher, NameMapper* mapper);
+  SearchRenderer(zim::Search* search, NameMapper* mapper);
```
Yeah, that makes sense. Making `SearchRenderer` depend only on `SearchResultSet`, the only object its concerned about seems more logical and the correct thing to do. I will make the changes. Thanks for the good explanation :)

kelson42 commented:
@maneeshpm @mgautierfr This PR ahould be rebased ans reviwed again,

mgautierfr reviewed (changes requested):
This is far better.

Still some code to change but nothing fundamental or complex to change.

maneeshpm commented on src/library.cpp:
```diff
@@ -143,27 +143,35 @@ Book& Library::getBookByPath(const std::string& path)
-  try {
-    return m_readers.at(id);
-  } catch (std::out_of_range& e) {}
+  if (m_readers.count(id)) {
+    return m_readers[id];
+  } else if (m_archives.count(id)) {
+    auto reader = make_shared<Reader>(m_archives[id]);
+    m_readers[id] = reader;
+    return reader;
+  }
```
Is there an advantage of this technique? or is it better coding style?

mgautierfr replied:
###### `at(key)` vs `[key]` :
`[key]` is stricky as if the key doesn't exist in the map, it will be created (with a nullptr shared_ptr here).
`at(key)` will throw a exception if the key is not found.
Because `[key]` create the entry, we must search if the key exist before using it (what you do).

###### `at(key)` vs `count(key)`+`[key]` :
With `at(key)` you do only one lookup. Either you find it and you get the entry, or not.
With the `count(key)`+`[key]` you need to do two lookup. On for the count, and then another one the get the entry.

On top of that, there is some kind of different semantics :
With a count, you means "If there is a key in the map, do something (which is return the entry)."
With the try/at you means "I want to return the entry, handling the case there is no entry in the map."

The second is a bit more direct and carry better what is your intention.

maneeshpm commented on include/tools/stringTools.h:
```diff
@@ -41,6 +41,7 @@ std::string removeAccents(const std::string& text);
 void loadICUExternalTables();
 
 std::string urlEncode(const std::string& value, bool encodeReserved = false);
+std::vector<std::string> getTitleVariants(const std::string& title);
```
Aah right, my bad. I'll make the change.

maneeshpm commented on src/tools/archiveTools.cpp:
```diff
@@ -105,5 +105,13 @@ zim::Entry getFinalEntry(const zim::Archive* const archive, const zim::Entry& en
   return archive->getEntryByPath(entry.getItem(true).getPath());
 }
 
+zim::Entry getEntryFromPath(const zim::Archive* const archive, const std::string& path)
+{
+  if (path.empty() || path == "/") {
+    return archive->getMainEntry();
+  }
+  return archive->getEntryByPath(path);
+}
+
```
I don't think we should add it to libzim because this is just a special user(libkiwix) need to get the `mainentry` if the requested entry is not found. Like you mentioned, there can be cases which as a library should not bother libzim. 

The new behaviour you suggested sounds good, I will make the change.

maneeshpm commented:
@kelson42 Rebase done. @mgautierfr The last three commits are the new changes.

mgautierfr commented on src/tools/archiveTools.cpp:
```diff
@@ -80,37 +80,41 @@ bool getArchiveFavicon(const zim::Archive* const archive,
+  try {
+    return archive.getEntryByPath(path);
+  } catch (zim::EntryNotFound& e) {
+    if (path.empty()) {
```
`path.empty() || path == "/"`

(I've told you that Reader behavior should not change, so use two different code path for reader and server.
But, I think we can share the new behavior, the change is pretty small and will not break existing usage of `Reader`)

mgautierfr reviewed (changes requested):
Just a small change to correctly handle `path=="/"` and we should be good.

kelson42 commented:
@mgautierfr Can you please review and have a look why the ci fails, this is since las week like this.

maneeshpm commented:
@kelson42 I reran the jobs, now the ci is passing. There is only the coverage issue, which we can expect since we have removed the usage of readers and searchers from internal server.

kelson42 commented:
@maneeshpm Nice, then you probably just need to write one or two additional tests and we are good

kelson42 commented:
I have rebase on master

maneeshpm commented:
@kelson42 I have added the tests and ci is passing now :smile: . 
Since we will be dropping more usage of `Reader` and `Searcher` from our codebase in the subsequent PRs, the `Reader` tests I added now somewhat demonstrates how its simply a wrapper on an `Archive` and how we can use an archive directly. I will add more tests in the coming PRs that will set us up for removing the wrappers completely.

mgautierfr commented on test/reader.cpp:
```diff
@@ -0,0 +1,62 @@
+    Reader reader("./test/zimfile.zim");
+    zim::Archive archive = *reader.getZimArchive();
+
+    std::ostringstream s;
+    s << archive.getUuid();
+
+    ASSERT_EQ(reader.getId(), s.str());
```
This show us that libzim API is not so nice.
We should add a cast to string to `zim::Uuid` to avoid this.

mgautierfr commented:
@maneeshpm Do you want to adapt the test now the libzim pr is merged ? (once the nightly is passed)
Or we merge now ? (it is not so important to change this as the test will be removed "soon")

maneeshpm commented:
@mgautierfr I expect the following PR to be a small one(and that depended on the libzim pr). So we can go ahead we the merge :D
